test/micro-kanren.ss

```diff
@@ -0,0 +1,322 @@
+;; Copyright 2019 Google LLC
+;;
+;; Licensed under the Apache License, Version 2.0 (the License);
+;; you may not use this file except in compliance with the License.
+;; You may obtain a copy of the License at
+;;
+;;     http://www.apache.org/licenses/LICENSE-2.0
+;;
+;; Unless required by applicable law or agreed to in writing, software
+;; distributed under the License is distributed on an AS IS BASIS,
+;; WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+;; See the License for the specific language governing permissions and
+;; limitations under the License.
+
+(library
+    (trivial)
+  (export do-test)
+  (import (rnrs))
+
+  ;;; The uKanren paper uses vectors to wrap variables. For now, we
+  ;;; tag pairs.
+  (define (var name)
+    `(var . ,name))
+
+  (define (var? x)
+    (and (pair? x) (eq? 'var (car x))))
+
+  (define (var=? x y)
+    (eq? (cdr x) (cdr y)))
+
+  ;;;; Unification
+  ;;; lookup is referred to as walk in the uKanren paper. The name
+  ;;; lookup is used in sokuza-karen.
+  (define (lookup var s)
+    (let ((binding (and (var? var) (assp (lambda (u) (var=? u var)) s))))
+      (if binding (lookup (cdr binding) s) var)))
+
+  ;;; A more diligent lookup.
+  (define (lookup* var s)
+    (let ((v (lookup var s)))
+      (cond
+       ((var? v) v)
+       ((pair? v) (cons (lookup* (car v) s) (lookup* (cdr v) s)))
+       (else v))))
+
+  (define (extend-s var val s)
+    `((,var . ,val) . ,s))
+
+  (define (unify u v s)
+    (let ((u (lookup u s)) (v (lookup v s)))
+      (cond
+       ((var? u) (extend-s u v s))
+       ((var? v) (extend-s v u s))
+       ((and (pair? u) (pair? v))
+        (let ((s (unify (car u) (car v) s)))
+          (and s (unify (cdr u) (cdr v) s))))
+       (else (and (eq? u v) s)))))
+
+  ;;;; Core
+  ;;; uKanren paper refers to these as unit and mzero.
+  (define (succeed s.c) `(,s.c))
+  (define (fail s.c) '())
+
+  (define (== u v)
+    (lambda (s.c)
+      (lambda ()
+        (let ((s (unify u v (car s.c))))
+          (if s (succeed `(,s . ,(cdr s.c))) (fail s))))))
+
+  ;;; Use int c in s.c pair to introduce a new variable to give to f
+  ;;; to build a goal.
```

Comment:
Similarly to `interleave` that replace `mplus`. It might be better to rename `s.c` to `substitutions+constant` or better (imo) `substitutions+index` (or even better move the constant / index to a stateful procedure `fresh-variable` it will allow to tremendously simplify the code and avoid the all around use of `car` to extract the substitution list).

At least, it might require a comment about the structure of `s.c` to explain how it is used in particular how the substitution list works in plain english (even if it is explained in `lookup` (and the paper)).

Reply:
?

Reply:
WDYT?

Reply:
Would the stateful procedure essentially be doing a `gensym`? I'd be a little concerned about adding too much state, because some of the elegance of uKanren is it's purely functional manner. Although, using state under the covers would make sense if we wanted to build this up to a full scale logic programming system.

I think adding `substitution+constant` in a comment wouldn't hurt. As far as names go though, I don't mind the short one. Out of curiosity, @jitwit, was there a problem parsing `s/c`, which I think is what the paper used? If so, we probably have a reader bug.

Reply:
No issues with the reader! I had just found the `s/c` name slightly confusing at first. Something broader like `state` might make more sense in case the representation changes or expands down the line.